pyth-sdk/src/price.rs

```diff
@@ -89,6 +90,148 @@ impl Price {
+
+        // get fractions for deposits
+        let deposits_as_price = Price {
+            price: utils::u64_to_i64(deposits)?,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let remaining_depositable_as_price = Price {
+            price: remaining_depositable,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let max_deposits_as_price = Price {
+            price: utils::u64_to_i64(max_deposits)?,
+            conf: 1,
+            expo: 0,
+            publish_time: 0,
+        };
+
+        let deposits_percentage = deposits_as_price.div(&max_deposits_as_price).ok_or(LiquidityOracleError::NoneEncountered)?;
+        let remaining_depositable_percentage = remaining_depositable_as_price.div(&max_deposits_as_price).ok_or(LiquidityOracleError::NoneEncountered)?;
+        
+        // get fractions for discount
+        let diff_discount_precision_initial_as_price = Price {
+            price: diff_discount_precision_initial,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let diff_discount_precision_final_as_price = Price {
+            price: diff_discount_precision_final,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let discount_precision_as_price = Price {
+            price: utils::u64_to_i64(discount_precision)?,
+            conf: 1,
+            expo: 0,
+            publish_time: 0,
+        };
+
+        let initial_percentage = diff_discount_precision_initial_as_price.div(&discount_precision_as_price).ok_or(LiquidityOracleError::NoneEncountered)?;
+        let final_percentage = diff_discount_precision_final_as_price.div(&discount_precision_as_price).ok_or(LiquidityOracleError::NoneEncountered)?;
+
+        // // compute left and right terms of the sum
+        let mut left = deposits_percentage.mul(&final_percentage).ok_or(LiquidityOracleError::NoneEncountered)?;
+        let mut right = remaining_depositable_percentage.mul(&initial_percentage).ok_or(LiquidityOracleError::NoneEncountered)?;
+
+        // scale left and right to match expo; need to ensure no overflow so have a match for NoneEncountered error
+        if left.expo > right.expo {
+            // prefer right scaled up to left if no overflow, to prevent any precision loss
+            let right_scaled = right.scale_to_exponent(left.expo);
+            let left_scaled = left.scale_to_exponent(right.expo);
+
+            match right_scaled {
+                Some(_x) => right = right_scaled.ok_or(LiquidityOracleError::NoneEncountered)?,
+
+                None => left = left_scaled.ok_or(LiquidityOracleError::NoneEncountered)?,
+            }
+        }
+        else if left.expo < right.expo {
+            // prefer left scaled up to right if no overflow, to prevent any precision loss
+            let left_scaled = left.scale_to_exponent(right.expo);
```

Comment:
I think scaling to a larger exponent is the exact operation that loses precision.